### PR TITLE
[TBAA] Do not rewrite TBAA if exists, always null out `!tbaa.struct`

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Retrieve `!tbaa` metadata via `!tbaa.struct` in `adjustForAccess` unless it already exists, as struct-path aware `MDNodes` emitted via `new-struct-path-tbaa` may be leveraged. As `!tbaa.struct` carries memcpy padding semantics among struct fields and `!tbaa` is already meant to aid to alias semantics, it should be possible to zero out `!tbaa.struct` once the memcpy has been simplified. `SROA/tbaa-struct.ll` test has gone out of scope, as `!tbaa` has already replaced `!tbaa.struct` in SROA.

Fixes regression: https://github.com/llvm/llvm-project/issues/95661.

### CODE REVIEW

**Comment by antoniofrighetto** [comment]
Kind ping. @nikic AFAICT I think this should basically reinstate behaviour prior to a87fa7f0caa1b8bf328524e0aec80a10ed4af7f9, modulo for taking into account scope/noalias too (as long as it was previously correct).

File: llvm/include/llvm/IR/Metadata.h [hunk 1]
```diff
@@ -846,8 +846,8 @@ struct AAMDNodes {
   AAMDNodes concat(const AAMDNodes &Other) const;
 
   /// Create a new AAMDNode for accessing \p AccessSize bytes of this AAMDNode.
-  /// If his AAMDNode has !tbaa.struct and \p AccessSize matches the size of the
-  /// field at offset 0, get the TBAA tag describing the accessed field.
+  /// If this AAMDNode has !tbaa.struct and \p AccessSize matches the size of
+  /// the field at offset 0, get the TBAA tag describing the accessed field.
```
Comment (fhahn) [inline llvm/include/llvm/IR/Metadata.h:850]: this should also mentioned that the existing tbaa metadata is preserve

**Comment by fhahn** [comment]
So the issue here is that clang generates legacy tbaa tags for tbaa.struct while the !tbaa tags are the struct-path format? Is there any reason that clang shouldn't emit struct-path tags for !tbaa.struct?

File: llvm/include/llvm/IR/Metadata.h (hunk 1 above)
Comment (antoniofrighetto) [reply llvm/include/llvm/IR/Metadata.h:850]: Adjusted, thanks.

**Comment by antoniofrighetto** [comment]
> Is there any reason that clang shouldn't emit struct-path tags for !tbaa.struct?

That seems to be the case, AFAICT, though this should be orthogonal wrt the above, if I'm not mistaken.

**Comment by fhahn** [comment]
> > Is there any reason that clang shouldn't emit struct-path tags for !tbaa.struct?
> 
> That seems to be the case, AFAICT, though this should be orthogonal wrt the above, if I'm not mistaken.

I am not sure, have inconsistent/conflicting `tbaa` and `tbaa.struct` nodes seems like it could cause other problems as well. If `tbaa.struct` nodes would be consistent with the `!tbaa` nodes, this patch shouldn't be needed?

**Comment by antoniofrighetto** [comment]
> > > Is there any reason that clang shouldn't emit struct-path tags for !tbaa.struct?
> > 
> > That seems to be the case, AFAICT, though this should be orthogonal wrt the above, if I'm not mistaken.
>
> I am not sure, have inconsistent/conflicting `tbaa` and `tbaa.struct` nodes seems like it could cause other problems as well. If `tbaa.struct` nodes would be consistent with the `!tbaa` nodes, this patch shouldn't be needed?

May there exist a specific reason why NewStructPathTBAA directly favours !tbaa metadata? This is at least what it seems to be from https://reviews.llvm.org/D41394. Maybe cc/ @kosarev?

**Comment by AaronBallman** [comment]
CC @rjmccall @efriedma-quic for an extra set of eyes on the changes.

**Comment by efriedma-quic** [comment]
Why do we have both tbaa and tbaa.struct on the same instruction in the first place?  That doesn't really seem useful for describing the aliasing; I'd think one or the other would be enough.  I guess the tbaa.struct allows describing the gaps?

If we do expect to have both tbaa and tbaa.struct available, in general, is there any point to trying to extract the tbaa info from the tbaa.struct?

If we fail to extract the field from the tbaa.struct, should we drop the tbaa.struct metadata?

**Comment by rjmccall** [comment]
Yeah, I thought the point of `tbaa.struct` was that it was necessary in different contexts.

**Comment by antoniofrighetto** [comment]
> Why do we have both tbaa and tbaa.struct on the same instruction in the first place? That doesn't really seem useful for describing the aliasing; I'd think one or the other would be enough. I guess the tbaa.struct allows describing the gaps?

Right, one describes the struct types for aliasing, the other seems to be used for padding:

https://github.com/llvm/llvm-project/blob/482c41e992c1edf8833a4577b56ff9dda49fbc83/clang/lib/CodeGen/CGExprAgg.cpp#L2201-L2211

In our example, for `t1.a = t2.a` assignment, we have the following:
```llvm
call void @llvm.memcpy.p0.p0.i64(ptr align 2 %a2, ptr align 2 %a1, i64 2, i1 false), !tbaa !4, !tbaa.struct !8
; ...
!4 = !{!5, !6, i64 0, i64 1}
!5 = !{!2, i64 4, !"_ZTS1T", !6, i64 0, i64 2, !6, i64 2, i64 2}
!6 = !{!2, i64 2, !"_ZTS7Wrapper", !7, i64 0, i64 2}
!7 = !{!2, i64 2, !"short"}
!8 = !{i64 0, i64 2, !9}
!9 = !{!7, !7, i64 0, i64 1}
```

> If we do expect to have both tbaa and tbaa.struct available, in general, is there any point to trying to extract the tbaa info from the tbaa.struct?

I don't believe so from what I can tell, avoiding extracting tbaa from tbaa.struct is indeed what this patch does.

> If we fail to extract the field from the tbaa.struct, should we drop the tbaa.struct metadata?

Intuitively, I think it could make sense. As far as  tbaa.struct is concerned, part of the users in the middle-end seem to be only transferring the metadata into the new memcpy / memmove. Not sure on this though.

**Comment by antoniofrighetto** [comment]
Earlier kind ping – in the hope to get this in before the branch, if it is ready.

**Comment by efriedma-quic** [comment]
> Right, one describes the struct types for aliasing, the other seems to be used for padding:

We should probably try to unify this at some point... but I guess we can put this aside for now.

> > If we fail to extract the field from the tbaa.struct, should we drop the tbaa.struct metadata?
> 
> Intuitively, I think it could make sense. As far as tbaa.struct is concerned, part of the users in the middle-end seem to be only transferring the metadata into the new memcpy / memmove. Not sure on this though.

I'd like to work this out now, while we're looking at when tbaa.struct needs to be nulled-out... what are the required invariants here?

**Comment by antoniofrighetto** [comment]
> > > If we fail to extract the field from the tbaa.struct, should we drop the tbaa.struct metadata?
> > 
> > 
> > Intuitively, I think it could make sense. As far as tbaa.struct is concerned, part of the users in the middle-end seem to be only transferring the metadata into the new memcpy / memmove. Not sure on this though.
> 
> I'd like to work this out now, while we're looking at when tbaa.struct needs to be nulled-out... what are the required invariants here?

Updated the patch to zero out !tbaa.struct when we fail to extract !tbaa from it (in addition to already zeroing it out when correctly retrieving it).

The preconditions are that aggregate assignments are turned into memcpy, and !tbaa.struct describes the pair offset-size of the types being copied. From what I can tell, this should not be needed anymore once the memcpy has been disaggregated, especially considering when we fail to extract !tbaa tag from it. I dropped `tbaa-struct.ll` test, as I think it went out of its original scope (a7929533300535547f8484f7e38765234a7e7c93) considering that !tbaa seems to be already leveraged (by SROA as well, no regression on llvm.org/pr42022).

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**